client/scala/src/main/scala/ai/verta/blobs/dataset/S3.scala

```diff
@@ -0,0 +1,168 @@
+) extends Dataset {
+  protected var contents = HashMap(metadataList: _*)
+  private var versionMap = HashMap(versionList: _*)
+
+  /** Get the version id of a file
+   *  @param path: S3 URL of a file in the form "s3://<bucketName>/<key>"
+   *  @return the version id of the file
+   */
+  def getVersionId(path: String) = versionMap.get(path)
+
+  override def equals(other: Any) = other match {
+    case other: S3 => contents.equals(other.contents) && versionMap.equals(other.versionMap)
+    case _ => false
+  }
+}
+
+/** Companion object to handle interaction with versioning blob */
+object S3 {
+  private val s3: AmazonS3 = AmazonS3ClientBuilder.standard().build()
+
+  /** Constructor that user should use:
+   */
+  def apply(paths: List[S3Location]): Try[S3] = {
+    val queryAttempt = Try(
+      paths.map(getS3LocMetadata).map(_.get)
+    ).map(_.flatten)
+
+    queryAttempt match {
+      case Failure(e) => Failure(e)
+      case Success(list) => Success(new S3(
+        list.map(metadata => metadata.path -> metadata),
+        list.filter(_.versionId.isDefined).map(metadata => metadata.path -> metadata.versionId.get)
+      ))
+    }
+  }
+
+  /** Factory method to convert a versioning blob instance
+   *  @param s3VersioningBlob the versioning blob to convert
+   */
+  def apply(s3VersioningBlob: VersioningS3DatasetBlob) {
+    var s3Blob = new S3(List(), List())
+    val componentList = s3VersioningBlob.components.get
+    val metadataList = componentList.map(
+      comp => comp.path.get.path.get -> s3Blob.toMetadata(comp.path.get)
+    )
+    val versionList = componentList.filter(_.s3_version_id.isDefined).map(
+      comp => comp.path.get.path.get -> comp.s3_version_id.get
+    )
+
+    s3Blob.contents = HashMap(metadataList: _*)
+    s3Blob.versionMap = HashMap(versionList: _*)
+
+    s3Blob
+  }
+
+  /** Convert a S3 instance to a VersioningBlob
+   *  @param blob S3 instance
+   *  @return equivalent VersioningBlob instance
+   */
+  def toVersioningBlob(blob: S3) = {
+    val s3Components = blob.components.map(comp => VersioningS3DatasetComponentBlob(
+      Some(comp),
+      blob.versionMap.get(comp.path.get)
+    ))
+
+    VersioningBlob(
+      dataset = Some(VersioningDatasetBlob(
+        s3 = Some(VersioningS3DatasetBlob(Some(s3Components)))
+      ))
+    )
+  }
+
+  /** Helper function to query metadata of S3 Location
+   */
+  private def getS3LocMetadata(loc: S3Location) = {
+    if (loc.key.isEmpty || loc.key.get.endsWith("/")) {
+      // no key (bucket), or is a folder
+      val versionListing =
+        if (loc.key.isEmpty) s3.listVersions(loc.bucketName, null)
+        else s3.listVersions(loc.bucketName, loc.key.get)
+
+      handleVersionListing(versionListing, List())
+    }
+    else {
+      val request =
+        if (loc.versionID.isDefined) new GetObjectMetadataRequest(loc.bucketName, loc.key.get, loc.versionID.get)
+        else new GetObjectMetadataRequest(loc.bucketName, loc.key.get)
```

Comment:
Can we collapse these?

Reply:
Hmm these are two different constructors (with and without versionId). I don't think there's a shorter way to branch here?

Reply:
Interesting. Ok.